### PR TITLE
Sharapov Georgiy. Lab 2.

### PR DESCRIPTION
**Pull request by nol0n**
Added a pass for opt that adds start_loop()/end_loop() calls before the start and in the end of loops.

### CODE REVIEW

File: llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        ParentModule->getOrInsertFunction("loop_end", funcType);
+
+    llvm::LoopAnalysis::Result &LI = FAM.getResult<llvm::LoopAnalysis>(F);
+    for (auto *L : LI) {
+      insertLoopStartEnd(L, loopStart, loopEnd);
+    }
+
+    auto PA = llvm::PreservedAnalyses::all();
+    PA.abandon<llvm::LoopAnalysis>();
+    PA.abandon<llvm::DominatorTreeAnalysis>();
```
Comment (m-ly4) [inline llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp]: Why do you think it should be abandoned too?

File: llvm/test/lab2/sharapov_georgiy/test.ll [hunk 2]
```diff
@@ -0,0 +1,190 @@
+; RUN: opt -load-pass-plugin=%llvmshlibdir/LoopFramer%shlibext -passes=LoopFramer -S %s | FileCheck %s
```
Comment (m-ly4) [inline llvm/test/lab2/sharapov_georgiy/test.ll]: Could you please also provide a test for a function with just `if-else` condition to check that no guards is added?

File: llvm/test/lab2/sharapov_georgiy/test.ll [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+define dso_local i32 @while_func() #0 {
```
Comment (m-ly4) [inline llvm/test/lab2/sharapov_georgiy/test.ll]: Could you also check that there is no additional loop guards inclusion if they are present there already?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp (hunk 1 above)
Comment (nol0n) [reply llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp]: I thought that if we add calls to some functions, then we are changing the control flow and, consequently, we can affect dominance of blocks.

Comment (m-ly4) [reply llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp]: Dominance can be changed if we add/move/delete blocks. Calls don't change blocks in caller function.

File: llvm/test/lab2/sharapov_georgiy/insert_test.ll [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+  ; CHECK: for.end:
+  ; CHECK-NEXT: call void @loop_end()
+for.end:                                          
+  %3 = load i32, ptr %b, align 4
+  ret i32 %3
+}
+
+
+define dso_local i32 @if_else_func(i32 noundef %b) #0 {
```
Comment (m-ly4) [inline llvm/test/lab2/sharapov_georgiy/insert_test.ll:200]: Better to write checks in the following manner:
`CHECK-LABEL: @if_else_func
CHECK-NOT: call void @loop_`

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**